opentracing-api/src/span.rs

```diff
@@ -0,0 +1,83 @@
+use SpanContext;
+use TagValue;
+
+/// The `Span` represents the OpenTracing specification's Span contract.
+pub trait Span<'a> {
+    /// The associated `SpanContext`.
+    type Context: SpanContext<'a>;
+
+    /// Retrieve the associated `SpanContext`.
+    ///
+    /// This may be called any time, including after `finish`.
+    fn context(&self) -> &Self::Context;
+
+    /// Sets a key:value tag on the `Span`.
+    fn set_tag<S>(&mut self, key: S, value: TagValue)
+    where
+        S: Into<String>;
+
+    /// Allows to unset a tag based on the given key. Noop if
+    /// it doesn't exist.
+    fn unset_tag<S>(&mut self, key: S)
+    where
+        S: Into<String>;
+
+    /// Returns a tag value if set, none otherwise
+    fn tag<S>(&self, key: S) -> Option<&TagValue>
+    where
+        S: Into<String>;
+
+    /// Record an event at the current walltime timestamp.
+    fn log(&mut self, event: String);
```

Comment:
With regards to the `event` type: what if it was any `serde::Serialize` type?
```rust
fn log<E>(&mut self, event: E)
where
    E: serde::Serialize;
```

Alternatively we can do what [`slog`](https://docs.rs/slog/2.2.1/slog/trait.Value.html) does and define an `Event` trait with an interface tailored to use by tracers.
Common types in the standard library would get an implementation of `Event` and users can implement it for any other type then need to log.